### ISSUE TITLE
sf::Music assertion failure when using Singleton

### ISSUE THREAD

**Issue by danieljpetersen** [SFML/SFML#3146]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I am here to report an issue and not to just ask a question or look for help (use the [forum](https://en.sfml-dev.org/forums/index.php#c3) or [Discord](https://discord.gg/nr4X7Fh) instead)

### Describe your issue here

> static void sf::priv::AudioDevice::unregisterResource(ResourceEntryIter): Assertion `instance && "AudioDevice instance should exist when calling AudioDevice::unregisterResource"' failed.

Introduced in this commit.

https://github.com/SFML/SFML/commit/cede2def8bd74f5bffd3044d27fcceccaaae56c2

### Your Environment

Linux, GCC 13.1

### Steps to reproduce

```
#include <SFML/Audio.hpp>

template <typename T>
class Singleton {
public:
	static T& the()
	{
		static T instance;
		return instance;
	}
};


class Test : public Singleton<Test> {
public:
	std::unique_ptr<sf::Music> music;
};

int main()
{
	Test::the(); // no issue if this line is commented out

	Test::the().music = std::make_unique<sf::Music>(sf::Music::openFromFile("input.ogg").value());

	// Test::the().music.release(); // no issue if this line is uncommented

	return 0;
}
```

### Expected behavior

.

### Actual behavior

.

**Comment by binary1248** [comment]
The reason this is happening is because this exact code construction manages to swap the ordering of destruction of the `AudioDevice` and the `sf::Music`. The assumption made in cede2def8bd74f5bffd3044d27fcceccaaae56c2 was that the order of destruction would always be the opposite of construction but your code demonstrates this doesn't always have to be the case. I can't think of any other solution than to revert cede2def8bd74f5bffd3044d27fcceccaaae56c2 if we want to keep supporting the use case you described here.

**Comment by danieljpetersen** [comment]
I guess the question is, what is the tangible benefit of the commit? It looks like it was primarily an optimization? I have a hard time imagining the cost of reference counting being meaningful here.

**Comment by binary1248** [comment]
I had my doubts as well in #3089, but in the end we ended up giving it a shot. I guess it turns out it didn't really work out as planned.

**Comment by vittorioromeo** [comment]
Surely #3147 must be reverted to support these particular use cases, however I am curious about the actual "real" code that caused this @danieljpetersen -- seems like the lifetime management could be simplified quite a bit, not sure why you'd ever need a `sf::Music` inside a `std::unique_ptr` inside a singleton...

**Comment by danieljpetersen** [comment]
I'm not sure why you felt the need to put real code in quotes. I'm also not sure why you felt the need to comment and chastise me for this given that you admit that we must revert, and also given that you quite publicly left the project.

I had an old project that I wanted to compile. It stored music on a singleton. I am not the only person that I know of who stored some SFML objects on singletons. I wrapped it in a unique pointer because that's how I chose to solve the issue in other contexts, and I wanted to deal with the removal of default constructors in a consistent manner. Wrapping them in a unique pointer seemed the least offensive way of doing so.

While I'm at it, I want to once again reiterate how unacceptable that change was to me. I've been using SFML for more than a decade, I root for and love this lib. I've now migrated away from SFML because of that change. I was trying to not come across as unhinged in my issue about that, but I perhaps underplayed how much I disliked it.

Cheers

**Comment by vittorioromeo** [comment]
I didn't put "real" in quotes to mock it, I put the quotes to emphasize that I wanted to see the *real* code and not a simplified example. 

I did not mean to chastize you at all -- I apologize if you felt that way. What I wanted is to provide advice on simplifying the lifetime hierarchy of your resources as the snippet you posted seemed overly complex.

I also don't mind that the change needed to be reverted, 3.x is work-in-progress and trying things out means that sometimes they do not work out, reverting is a totally fine decision in this case if these use cases need to be supported.

Leaving the project also doesn't mean I don't want to be involved with it in any way or form in the future.

> I wanted to deal with the removal of default constructors in a consistent manner. Wrapping them in a unique pointer seemed the least offensive way of doing so

I will reiterate, as mentioned in https://github.com/SFML/SFML/issues/3120#issuecomment-2184263169, that it is not necessary to reintroduce empty states via unique pointers or `std::optional`, and in fact that choice will make your user experience much worse than it needs to be -- the `sf::Music` doesn't need to be wrapped in any additional layer, it can be initialized through the constructor.

```cpp
struct Test : public Singleton<Test> 
{
    sf::Music music = sf::Music::openFromFile("input.ogg").value();
};
```

I was interested in seeing the *real* code to understand if the `std::unique_ptr` was somehow really necessary, or not.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/compare/master...audio_construction_order_crash

I'm trying to write a test to recreate this crash but can't get it to fail. Did I not properly copy your example code?

**Comment by vittorioromeo** [comment]
> [master...audio_construction_order_crash](https://github.com/SFML/SFML/compare/master...audio_construction_order_crash)
> 
> I'm trying to write a test to recreate this crash but can't get it to fail. Did I not properly copy your example code?

@ChrisThrasher: I can reproduce the issue with this code:

```cpp
#include <SFML/Audio/Music.hpp>

#include <iostream>
#include <optional>

template <typename T>
class Singleton
{
public:
    static T& getInstance()
    {
        std::cout << "Singleton<T>::getInstance()\n";

        static T instance;
        return instance;
    }
};

struct MusicSingleton : Singleton<MusicSingleton>
{
    MusicSingleton()
    {
        std::cout << "MusicSingleton::MusicSingleton()\n";
    }

    std::optional<sf::Music> music;
};

int main()
{
    std::cout << "-- start of main --\n\n";
    MusicSingleton::getInstance();
    std::cout << "\n-- after first getInstance --\n\n";
    MusicSingleton::getInstance().music = sf::Music::openFromFile("resources/ding.flac").value();
    std::cout << "\n-- end of main-- \n\n";
}
```

Adding some printing to the various pieces involved explains why it fails:

<table>
<tr>
<th align="center">
<img width="441" height="1px">
<p> 
<small>
Original example
</small>
</p>
</th>
<th align="center">
<img width="441" height="1">
<p> 
<small>
Removing first `MusicSingleton::getInstance();` call
</small>
</p>
</th>
</tr>
<tr>
<td>

```text
-- start of main --

Singleton<T>::getInstance()
MusicSingleton::MusicSingleton()

-- after first getInstance --

Music::openFromFile(...)
::getDeviceState()
DeviceState::DeviceState()
AudioResource::AudioResource() | rc = 1
AudioDevice::AudioDevice()
Music::Music(...)
::getDeviceState()
AudioResource::AudioResource() | rc = 2
Music::~Music()
::getDeviceState()
AudioResource::~AudioResource() | rc = 1
Singleton<T>::getInstance()
::getDeviceState()
AudioResource::AudioResource() | rc = 2
Music::~Music()
::getDeviceState()
AudioResource::~AudioResource() | rc = 1

-- end of main--

DeviceState::~DeviceState()
AudioDevice::~AudioDevice()
Music::~Music()
~SoundStream::Impl()
MiniaudioUtils::SoundBase::~SoundBase()
AudioDevice::unregisterResource(...)

(!) Assertion failed: "..."
```
  
</td>
<td>
  
```text
-- start of main --

-- after first getInstance --

Music::openFromFile(...)
::getDeviceState()
DeviceState::DeviceState()
AudioResource::AudioResource() | rc = 1
AudioDevice::AudioDevice()
Music::Music(...)
::getDeviceState()
AudioResource::AudioResource() | rc = 2
Music::~Music()
::getDeviceState()
AudioResource::~AudioResource() | rc = 1
Singleton<T>::getInstance()
MusicSingleton::MusicSingleton()
::getDeviceState()
AudioResource::AudioResource() | rc = 2
Music::~Music()
::getDeviceState()
AudioResource::~AudioResource() | rc = 1

-- end of main--

Music::~Music()
~SoundStream::Impl()
MiniaudioUtils::SoundBase::~SoundBase()
AudioDevice::unregisterResource(...)
::getDeviceState()
AudioResource::~AudioResource() | rc = 0
AudioDevice::~AudioDevice()
DeviceState::~DeviceState()

-- all is well -- 
```
  
</td>

</table>

Basically, the first `Test::the();` call ends up putting the music singleton on top of the stack of globals to be destroyed. When the music is created next, the `AudioDevice` is initialized, which is put on top of the globals stack.

When the stack is resolved, the `AudioDevice` gets destroyed before the music, causing the assertion failure.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/commit/8277fe17101c8608918b1a2757f828bca292b18e

I was able to fix my problem by putting these new tests at the top of the `TEST_CASE` so they're ran before any other music objects are initialized.
